### ISSUE TITLE
OSRM-routed for Production Setup

### ISSUE THREAD

**Issue by rajat23495** [Project-OSRM/osrm-backend#5744]
Currently, i have a production ready setup of OSRM(nodejs bindings) in ECS fargate based cluster where my map data lies on an EFS shared across multiple containers. On issue that we are facing is /health URL seems to be working fine but when table or route service request is made, it takes around 50s to yield a response.
Map - Planet data
Size specs - 16GB RAM, 2vcpu
My hypothesis - Since health check does not require data osrm data files, it is not failing but for table/route services, data needs to be loaded into memory for performance. Is there any way where i can make my system more efficient? AM i missing something here?

**Comment by danpat** [comment]
Are you using the `mmap` option?  If so, you can expect extremely slow response times using EFS until the data is warmed up in the filesystem cache.

**Comment by rajat23495** [comment]
We are not using mmap options in nodejs wrapper for calling libosrm. I have a very basci doubt actually, if we say its a memory intensive process, what memory utilisation should i expect in my metrics. I have configured this service on AWS fargate as well as EC2 based cluster. In EC2 based cluster althought he machine size is 64GB but as per the graph memory utilisation max is only around 25-30%.

**Comment by danpat** [comment]
OSRM has several different modes of operation w.r.t. data loading and memory access.  Can you paste the code where you initialize the OSRM object in NodeJS?  Also, what are the size of the files you have?  By default, OSRM would load all of them except the `.fileIndex` file into RAM, so you would need roughly that amount of RAM.

50 seconds is completely unreasonable, something is very wrong with your setup.  A properly configured OSRM server calculates most routes in more like 5-10ms.

**Comment by rajat23495** [comment]
`const os = require("os");

process.env.UV_THREADPOOL_SIZE = Math.ceil(os.cpus().length * 1.5);

const OSRM = require("osrm");

function loadGraph(options) {
  const opts = { path: options.osrmDataPath, algorithm: "MLD" };
  return new OSRM(opts);
}

module.exports = {
  loadGraph
};`

`const http = require("http");
const express = require("express");
const bodyParser = require("body-parser");
const helmet = require("helmet");
const logfmt = require("logfmt");

const carhealth = require("./routes/health");
const carroute = require("./routes/route");
const cartable = require("./routes/table");
const carnearest = require("./routes/nearest");

const osrmBindings = require("./lib/osrm");

function configureMiddlewares(app) {
  app.use(helmet());
  app.use(bodyParser.json());
  app.use(bodyParser.urlencoded({ extended: false }));
  app.use(logfmt.requestLogger());
}

function configureRoutes(app) {
  app.use("/car/health", carhealth);
  app.use("/car/route", carroute);
  app.use("/car/table", cartable);
  app.use("/car/nearest", carnearest);
}

function configureOSRM(app, options) {
  const osrm = osrmBindings.loadGraph(options);
  app.set("osrm", osrm);
}

function createServer(options) {
  const opts = options || {};
  const app = express();
  configureOSRM(app, opts);
  configureMiddlewares(app);
  configureRoutes(app);
  return http.createServer(app);
}

module.exports = {
  createServer
};`

`const express = require("express");
const logfmt = require("logfmt");

const router = express.Router();

router.post("/", (req, res) => {
  if (!req.body.coordinates) {
    return res.status(422).json({ error: "Missing coordinates" });
  }

  const osrm = req.app.get("osrm");
  const options = {
    coordinates: req.body.coordinates,
    sources: req.body.sources,
    destinations: req.body.destinations,
    exclude: req.body.exclude,
    annotations: ['duration','distance']
  };

  if (!req.body.sources || !req.body.destinations) {
    delete options.sources;
    delete options.destinations;
  }
  if (!req.body.exclude) {
    delete options.exclude;
  }

  try {
    osrm.table(options, (err, result) => {
      if (err) {
        return res.status(422).json({ error: err.message });
      }
      return res.json(result);
    });
  } catch (err) {
    logfmt.error(new Error(err.message));
    return res.status(500).json({ error: err.message });
  }
});

module.exports = router;`

Attached afew code snippets here for reference.

**Comment by danpat** [comment]
@rajat23495  that code looks reasonable to me.  I was looking for things like accidentally re initializing OSRM each time a request came in, but the flow looks fine, you only initialize OSRM once.

I haven't used Fargate myself - are you sure it's not doing something dumb like firing up new containers as each request comes in?

**Comment by rajat23495** [comment]
@danpat the table requests are pretty small actually, we generate only a 2x2 matrix. That too for distances. Fargate recently introduced integration with AWS EFS so thats why we thought of moving to fargate but after performance testing we doubt we will be able to migrate to fargate. Probably whaty i feel is files are not loaded into memory and so the response time seem to be very high and as u said it might be possibly swapping a lot. Not sure though.
Attaching a request snippet - 
```
{
	"coordinates" : [[87.2972029,12.7890654],[98.667244,12.8905055],[77.683082, 12.913107],[77.6974825,12.9231607],[77.7044041,13.1986348]],
	"annotations" : "distance,duration",
	"sources" :  [0],
	"destinations": [1]
}
```

**Comment by danpat** [comment]
Yeah, given all the above, I'd be asking questions about real RAM on your fargate instances - if your routing data is in RAM (not all 280GB of those files is needed, it's more like 30-40GB for the actual routing files), then a 2x2 matrix prepared using CH should take just a couple of milliseconds.

It doesn't seem like it should be *impossible* to use Fargate here - have you tried running a warm-up process before sending in some requests?  Perhaps there is some lazy initialization going on that means the first few requests are very slow, then they become fast after that?

**Comment by rajat23495** [comment]
@danpat yes actually i tried a few requests initially i did notice this behaviour but whenever i change lat longs, response time is high probably because those set of coordinates might have not been loaded into memory. 
Even i feel fargate seems to be a good option because of less management overhead.

**Comment by venkatavendra** [comment]
@danpat do you have any suggestions regarding heap size in this nodejs application as @rajat23495 mentioned we are using 64 memory & 8 cpu EC2 machine being used here. Recently  facing GC issues. Heap size is not mentioned so the default value is 700mb.  So if I want increase what is your recommendation? 

Right now I changing it to 8GB memory for heap size.

As in

**Comment by venkatavendra** [comment]
@danpat as you suggested for @rajat23495  

do a warm-up process before sending in some requests?
I tried to warm-up by using dd command in docker container which contains 64 gb memory and car profile is around 50GB memory
"dd if=/car/planet-latest.osrm of=/dev/mem" (getting an insufficient memory)"
Are there any other approaches or am I doing wrong?


Perhaps there is some lazy initialization going on that means the first few requests are very slow, then they become fast after that? 
- Yes, it looks like that. In my testing, I am seeing first requests are very slow and then become fast after that. 

> actually not slow also getting 504 responses and I increased timeouts in ALB's and nodejs timeouts still not working

**Comment by danpat** [comment]
`dd if=/car/planet-latest.osrm of=/dev/mem` - that's definitely not what you want to do. Use `of=/dev/null` instead - you want the kernel page cache to get warmed up, not overwrite all your operating memory blindly.  Writing to `/dev/mem` will almost never be a thing you'd want to do directly....

What you want to do is have the kernel *read* all of `/car/planet-latest.osrm` so it's in the filesystem cache - you don't need to *write* it to anywhere in particular, you just need future reads of the fail (which OSRM would be doing) to be fast.

**Comment by venkatavendra** [comment]
@danpat my first attempt was  dd if=/car/planet-latest.osrm of=/dev/null. It is not working initial requests still very slow and getting 504's.

And then tried dd if=/car/planet-latest.osrm of=/dev/mem & dd if=/car/planet-latest.osrm of=/dev/urandom

thanks, I will check on this mmap option to osrm-routed, or the NodeJS bindings are we using or not.
